server/src/main/java/org/elasticsearch/common/document/DocumentField.java

```diff
@@ -66,7 +66,12 @@ public DocumentField(String name, List<Object> values, List<Object> ignoredValue
 
     public DocumentField(String name, List<Object> values, List<Object> ignoredValues, List<LookupField> lookupFields) {
         this.name = Objects.requireNonNull(name, "name must not be null");
-        this.values = Objects.requireNonNull(values, "values must not be null");
+        if (values == null) {
+            throw new NullPointerException("values must not be null");
+        } else {
+            // copy to avoid field value cache clear
+            this.values = List.copyOf(values);
```

Comment:
I would prefer to do the value copying in `StoredValueFetcher#fetchValues()`  which is the one responsible for the "_id" field. That way we avoid copying other fields lists values that are not cleared with the cached field data in LeafStoredFieldsLookup.